core/server/master/src/main/java/alluxio/master/file/replication/ReplicationChecker.java

```diff
@@ -131,13 +137,101 @@ public void heartbeat() throws InterruptedException {
     // Check the set of files that could possibly be over-replicated
     inodes = mInodeTree.getReplicationLimitedFileIds();
     check(inodes, mReplicationHandler, Mode.EVICT);
+
+    // Check the set of files that could possibly be mis-replicated
+    inodes = mInodeTree.getPinIdSet();
+    checkMisreplicated(inodes, mReplicationHandler);
   }
 
   @Override
   public void close() {
     // Nothing to clean up
   }
 
+  /**
+   * Find a set of from and to locations for our moveBlock operation.
+   * After the move, the file will be in the desired pinned medium at least minReplication times.
+   *
+   * @param file the file to scan for misplaced blocks
+   * @param blockInfo blockInfo containing information about the block locations
+   * @return a map that maps from workerHost to desired medium
+   */
+  private Map<String, String> findMisplacedBlock(
+      InodeFile file, BlockInfo blockInfo) {
+    Set<String> pinnedMediumTypes = file.getMediumTypes();
+    Map<String, String> movement = new HashMap<>();
+    if (pinnedMediumTypes.isEmpty()) {
+      // nothing needs to be moved
+      return Collections.emptyMap();
+    }
+    // at least pinned to one medium type
+    String firstPinnedMedium = pinnedMediumTypes.iterator().next();
+    int minReplication = file.getReplicationMin();
+    int correctReplication = 0;
+    List<String> candidates = new ArrayList<>();
+    for (BlockLocation loc: blockInfo.getLocations()) {
+      if (loc.getMediumType().equals(firstPinnedMedium)) {
```

Comment:
Why do we check this condition instead of `pinnedMediumTypes.contains(loc.getMediumType())`?

Reply:
because we are not address multiple medium types being pinned right now. but yeah I agree, your solution covers both cases.